### PR TITLE
WIP: Import from FEniCS and simplify code.

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**

### CODE REVIEW

File: partitioned-heat-conduction/fenics/utils/utils.py [hunk 1]
```diff
@@ -13,20 +13,13 @@ def b_splines(precice, degree, dt):
     '''
     # use equidistant samples
     # apparently you need 2k+2 samples for degree k
-    x_dist = float(dt) / (2 * degree + 2)
-    nodes = []
+    nodes = np.linspace(0,dt,2*3+3)  # <-- But these are 2k+3 samples
```
Comment (BenjaminRodenberg) [inline partitioned-heat-conduction/fenics/utils/utils.py]: This contradicts the statement above. Which one is true?

File: partitioned-heat-conduction/fenics/heat.py [hunk 2]
```diff
@@ -160,14 +159,14 @@ def determine_gradient(V_g, u, flux):
 bc = []
 # Create for each dimension of Vbig a coupling expression for either the time derivatives (Dirichlet side)
 #  or Neumann side (no time derivatives required as they are enforced with changing F)
-coupling_expressions = [precice.create_coupling_expression()] * tsm.num_stages
+coupling_expressions = [precice.create_coupling_expression() for _ in range(tsm.num_stages)]
```
Comment (BenjaminRodenberg) [inline partitioned-heat-conduction/fenics/heat.py:162]: This was definitely a nasty bug: `[something] * N` creates a list with `N` references to the same object. You can check this by printing `coupling_expressions` and looking at the names of the expressions.

Still: I do not get the correct solution, but this should be a step into the right direction.

**Review by BenjaminRodenberg — COMMENTED** [review]
Noteworthy observation: If I try to use a different time stepping scheme via `tsm = BackwardEuler()` or `tsm = RadauIIA(1)` I get an error. Nothing we have to fix now, but important to know that we cannot use arbitrary time stepping methods for some reason. @NiklasVin do you observe the same problem? Maybe this is also due to my FEniCS installation and some compatibility problem.

File: partitioned-heat-conduction/fenics/heat.py (hunk 2 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction/fenics/heat.py:162]: This bug was actually hiding another bug with the Neumann BCs. By fixing this bug the Neumann BC bug became visible. This also caused the (previously working) Fixme 4 part to fail. See below for comment on the Neumann BC bugfix.

File: partitioned-heat-conduction/fenics/heat.py [hunk 3]
```diff
@@ -160,28 +159,26 @@ def determine_gradient(V_g, u, flux):
...
-            bc. append(DirichletBC(Vbig.sub(i), coupling_expressions[i], coupling_boundary))
-            # Fixme 4: Damit würde man die richtigen Ergebnisse erhalten
-            # bc.append(DirichletBC(Vbig.sub(i), du_dt[i], coupling_boundary))
+            bc.append(DirichletBC(Vbig.sub(i), coupling_expressions[i], coupling_boundary))
     else:
         vs = split(v)
         for i in range(tsm.num_stages):
             bc.append(DirichletBC(Vbig.sub(i), du_dt[i], remaining_boundary))
-            F += vs[i] * coupling_expressions[i] * dolfin.ds
+            F += vs[i] * coupling_expressions[-1] * dolfin.ds
```
Comment (BenjaminRodenberg) [inline partitioned-heat-conduction/fenics/heat.py]: It looks like we need to use the heat flux corresponding to the end of the current time step for the Neumann boundary condition. I currently do not have a good explanation why this is the case, but if we think about the way how Neumann boundary conditions work in general and compare this to the way Dirichlet work it intuitively makes sense to me. However, there is certainly a better explanation including more hard facts and less intuition :smirk: 

This problem was masked by the bug in the creation of the coupling expressions because due to the pointer-nature of the problem, all coupling expressions were pointing to the coupling expressions corresponding to the end of the current time step.

File: partitioned-heat-conduction/fenics/heat.py [hunk 4]
```diff
@@ -160,28 +159,26 @@ def determine_gradient(V_g, u, flux):
...
 else:
     if problem is ProblemType.DIRICHLET:
         bc.append(DirichletBC(Vbig, du_dt[0], remaining_boundary))
         bc.append(DirichletBC(Vbig, coupling_expressions[0], coupling_boundary))
     else:
         bc.append(DirichletBC(Vbig, du_dt[0], remaining_boundary))
-        F += v * coupling_expressions[0] * dolfin.ds
+        F += v * coupling_expressions[-1] * dolfin.ds
```
Comment (BenjaminRodenberg) [inline partitioned-heat-conduction/fenics/heat.py]: See above.

**Review by BenjaminRodenberg — COMMENTED** [review]
I think I found the problem :tada:

File: partitioned-heat-conduction/fenics/heat.py (hunk 3 above)
Comment (BenjaminRodenberg) [reply partitioned-heat-conduction/fenics/heat.py]: Another thing we should keep in mind here: If we are using a time stepping scheme, where `c[-1]` does not correspond to the end of the window, things will probably become tricky. Instead of referring to the `coupling_expression` via the index, I think we need a different strategy for creating `coupling_expression`s for the Neumann participant. What's also important: `coupling_expression[0]` is not needed for the Neumann participant. So why bother creating it in the first place?

Comment (BenjaminRodenberg) [reply partitioned-heat-conduction/fenics/heat.py]: Maybe I was a bit quick here: One thing we should definitely test is what happens, if the heat flux changes over time. Currently, we have a constant heat flux and this might hide some possible errors.

In d410f66abafd372031f036357b96a03f00aa7201 I realized that the heat flux was not initialized (no problem with implicit Euler). This results in zero heat flux for `t=0`, which also leads to an incorrect interpolant. I think this is the much more reasonable fix.

File: partitioned-heat-conduction/fenics/utils/utils.py (hunk 1 above)
Comment (NiklasVin) [reply partitioned-heat-conduction/fenics/utils/utils.py]: The comment about the number of samples needed was a bit imprecise. You need  _at least_  2k+2 samples.
I will check, if you can reduce the samples to at least 2k, as I changed this code segment here:
[https://github.com/NiklasVin/tutorials/commit/00007d7be9a4d8334089c2f91b3a25e49ed6b0ec#diff-0b22c9acd2661e0e0fe1d0350fdd7736431cd134ab6f1e5b088d90b0f8c7b2b8L34-R35]()
If you used the constructor directly instead of using the return values of `splrep`, the constructor had for some reason a problem less than 2k+2 samples were given. Maybe an implementation error from my side, but I am on it :)

File: partitioned-heat-conduction/fenics/heat.py [hunk 5]
```diff
@@ -181,7 +178,7 @@ def determine_gradient(V_g, u, flux):
         bc.append(DirichletBC(Vbig, coupling_expressions[0], coupling_boundary))
     else:
         bc.append(DirichletBC(Vbig, du_dt[0], remaining_boundary))
-        F += v * coupling_expressions[0] * dolfin.ds
+        F += v * coupling_expressions[i] * dolfin.ds
```
Comment (BenjaminRodenberg) [inline partitioned-heat-conduction/fenics/heat.py]:
```suggestion
        F += v * coupling_expressions[0] * dolfin.ds
```

**Comment by NiklasVin** [comment]
> Noteworthy observation: If I try to use a different time stepping scheme via `tsm = BackwardEuler()` or `tsm = RadauIIA(1)` I get an error. 

Is this still valid for the most recent commit? Using [1b9d9cb](https://github.com/NiklasVin/tutorials/pull/1/commits/1b9d9cbda9dab59b5cab73bd41bde3cd738ef014) both time stepping schemes work fine for me.

File: partitioned-heat-conduction/fenics/heat.py (hunk 3 above)
Comment (NiklasVin) [reply partitioned-heat-conduction/fenics/heat.py]: > It looks like we need to use the heat flux corresponding to the end of the current time step for the Neumann boundary condition.

Yes, we need to set the heat flux to the end of the time step of the respective stages.
Otherwise it would e.g. not comply with the following statement from the [Irksome paper](https://doi.org/10.48550/arXiv.2006.16282): 

> We note that weakly-enforced boundary conditions [...] require [...] just the evaluation of any time-dependent data at the correct stage times.

So for the weak form for the stage $`k_i`$ at $`t+c_i\cdot\delta t`$.

I tested this as well with $`g_{tri}`$ (from the [Quasi-Newton waveform iteration]( https://doi.org/10.1002/nme.6443) paper) to check if the convergence of the solution is equal to the order of the time stepping scheme. With the introduction of a time dependent heat flux, wrongly imposed Neumann Boundaries would impact the convergence of the solution.
Until now, I tested it with LobattoIIIC(2) and the results look promising:
Neumann BC at $`t+c_i\cdot\delta t`$
| $`\Delta t`$  | $`\delta t`$ | L2 Dirichlet | L2 Neumann|
| ------------- | ------------ | ------------ | ------------- | 
| 0.1  | 0.05 | 1,15E-05 | 7,09E-05|
| 0.05  | 0.025 | 2,59E-06 | 1,74E-05|
| 0.025  | 0.0125 | 6,51E-07| 4,32E-06|

So the convergence order approximately is, as expected, two.

As I was not quite sure, if you referred to the time of each stage by "end of current time step" or the actual end of the current time step, and as a sanity check, if I understood the idea behind the k-ansatz correctly, I also tested, how the errors look like if you only use data at the boundaries retrieved only from $`t+\delta t`$. And, thankfully, the results have no convergence order of two (otherwise I would, again, be clueless why this is happening😅):
| $`\Delta t`$  | $`\delta t`$ | L2 Dirichlet | L2 Neumann|
| ------------- | ------------ | ------------ | ------------- | 
| 0.1  | 0.05 | 2,73E-04 | 3,98E-04
| 0.05  | 0.025 | 1,59E-04 | 2,08E-04
| 0.025  | 0.0125 | 8,96E-05| 1,07E-04

Comment (BenjaminRodenberg) [reply partitioned-heat-conduction/fenics/heat.py]: These are exactly the tests I meant with

> One thing we should definitely test is what happens, if the heat flux changes over time.

Thank you! Looks good.

Comment (NiklasVin) [reply partitioned-heat-conduction/fenics/heat.py]: Last but not least, a comment about:
>  If we are using a time stepping scheme, where c[-1] does not correspond to the end of the window, things will probably become tricky.

I think this should be fine. As 
> Yes, we need to set the heat flux to the end of the time step of the respective stages. Otherwise it would e.g. not comply with the following statement from the [Irksome paper](https://doi.org/10.48550/arXiv.2006.16282):
> 
> > We note that weakly-enforced boundary conditions [...] require [...] just the evaluation of any time-dependent data at the correct stage times.

Additionally, tested with the [GaussLegendre(1)](https://en.wikipedia.org/wiki/Gauss%E2%80%93Legendre_method), the time dependent heat flux and the Neumann BC only used at $`t+\frac{\delta t}{2}`$, the order seems fine. 
So it is nice to see, that actually no special treatment for such time stepping methods are required. I hope for the best, that this will also be true for multi step methods.

**Comment by BenjaminRodenberg** [comment]
> > Noteworthy observation: If I try to use a different time stepping scheme via `tsm = BackwardEuler()` or `tsm = RadauIIA(1)` I get an error.
> 
> Is this still valid for the most recent commit? Using [1b9d9cb](https://github.com/NiklasVin/tutorials/pull/1/commits/1b9d9cbda9dab59b5cab73bd41bde3cd738ef014) both time stepping schemes work fine for me.

Yes. I still get the following error:

```
Traceback (most recent call last):
  File "heat.py", line 108, in <module>
    tsm = BackwardEuler()
  File "/home/benjamin/Programming/tutorials/partitioned-heat-conduction/fenics/utils/ButcherTableaux.py", line 137, in __init__
    super(BackwardEuler, self).__init__(1)
  File "/home/benjamin/Programming/tutorials/partitioned-heat-conduction/fenics/utils/ButcherTableaux.py", line 127, in __init__
    L = FIAT.GaussRadau(U, num_stages - 1)
AttributeError: module 'FIAT' has no attribute 'GaussRadau'
```

I think this is probably a problem with the specific version of FEniCS or one of the dependencies I'm using. I'm using the following versions:

```
pip3 list
...
fenics                       2019.1.0
fenics-basix                 0.4.2
fenics-dijitso               2019.1.0
fenics-dolfin                2019.2.0.dev0
fenics-dolfinx               0.4.1
fenics-ffc                   2019.1.0.post0
fenics-ffcx                  0.4.2
fenics-fiat                  2019.1.0
fenics-ufl                   2019.1.0
...
```

This might also be a problem with my local FEniCS installation. So let's not worry about this issue here. If the other time stepping schemes work on your side that's good. If we want to test things properly and exclude side effects originating from the runtime system we can use the docker image provided by the FEniCS project and make the time stepping scheme configurable through some command line argument. But this is not really relevant yet.
